Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,157 @@
+  -- in the image of the matrix. This changes @z@.
+  | potentialSolution == mempty && counter > 5  = linearSolveHelper 1 matrix otherVecs 1
+  -- This is a good solution.
+  | otherwise                                   = potentialSolution
+  where
+    potentialSolution = findSolution singularities matrix almostZeroVector
+    almostZeroVector = evaluate matrix z reducedMinPoly
+    (singularities, reducedMinPoly) = L.break unBit (U.toList $ unF2Poly potentialMinPoly)
+    -- Information gathered from the previous random vector @x@ is used in
+    -- the subsequent iteration.
+    potentialMinPoly = lcm previousPoly candidateMinPoly
+    candidateMinPoly = findCandidatePoly matrix z x
+
+-- Infinite lists of random DBVectors.
+getRandomDBVectors :: forall k. KnownNat k => Double -> StdGen -> [DBVector k]
+getRandomDBVectors density gen = go $ randomRs (0, 1) gen
+  where
+    numberOfColumns = naturalToInt $ natVal (Proxy :: Proxy k)
+    go :: KnownNat k => [Double] -> [DBVector k]
+    go list = newVector `seq` newVector : go backOfList
+      where
+        newVector = DBVector $ fromJust $ SU.fromList $ map (\d -> Bit (d < density)) frontOfList
+        (frontOfList, backOfList) = L.splitAt numberOfColumns list
+
+-- The input is a matrix @B@ and random vectors @z@ and @x@. It returns the
+-- sequence @x A ^ k z@ as k runs from 0 to the size of the matrix.
+generateData :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k -> F2Poly
+generateData matrix z x = toF2Poly $ U.fromList $ reverse $ map (x `dot`) matrixPowers
+  where
+    matrixPowers = take (((*2) . naturalToInt . natVal) matrix) $ L.iterate (matrix `mult`) z
+
+-- This routine finds the generating polynomial of the random sequence computed
+-- in @generateData@.
+berlekampMassey :: Int -> F2Poly -> F2Poly -> F2Poly
+berlekampMassey dim = go 1 0
+  where
+    go :: F2Poly -> F2Poly -> F2Poly ->F2Poly -> F2Poly
+    go oneBefore twoBefore a b
+      | U.length (unF2Poly b) <= dim = oneBefore
+      -- Updated value is given by @twoBefore - oneBefore * q@
+      | otherwise                    = go (twoBefore - oneBefore * q) oneBefore b r
+        where
+          (q, r) = quotRem a b
+
+findCandidatePoly :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k -> F2Poly
+findCandidatePoly matrix z x = berlekampMassey dim errorPoly randomSequence
+  where
+    randomSequence = generateData matrix z x
+    errorPoly = fromInteger (1 `shiftL` (2 * dim)) :: F2Poly
+    dim = naturalToInt $ natVal matrix
+
+-- This routine takes a polynomial @p@, a @matrix@ and a vector @w@ and
+-- returns @p(A)w@.
+evaluate :: KnownNat k => SBMatrix k -> DBVector k -> [Bit] -> DBVector k
+evaluate matrix w = foldr (\coeff acc -> (matrix `mult` acc) <> (if unBit coeff then w else mempty)) mempty
+
+-- Tries to infer a solution. If it does not succeed, it returns an empty solution.
+findSolution :: KnownNat k => [Bit] -> SBMatrix k -> DBVector k -> DBVector k
```

Comment:
This function never uses values of its first arguments, only its length. That said, let's pass the length of singularities prefix as `Int` instead of `Bit`.